[DataPipe] adding ZipperMapDataPipe

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68032

Part of #57031

cc @VitalyFedyunin @ejguan @NivekT

Differential Revision: [D32263058](https://our.internmc.facebook.com/intern/diff/D32263058)

NivekT committed:
[DataPipe] adding ZipperMapDataPipe

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] adding ZipperMapDataPipe"


Part of #57031

cc @VitalyFedyunin ejguan @NivekT

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32263058).

NivekT committed:
Update on "[DataPipe] adding ZipperMapDataPipe"


Part of #57031

cc @VitalyFedyunin ejguan @NivekT

Differential Revision: [D32263058](https://our.internmc.facebook.com/intern/diff/D32263058)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32263058).

ejguan commented on torch/utils/data/datapipes/map/combining.py:
```diff
@@ -43,3 +43,42 @@ def __len__(self) -> int:
+            try:
+                res.append(dp[index])
+            except IndexError:
+                raise IndexError(f"Index {index} is out of range because {dp} has a smaller length.")
```
nit: The error message may give wrong information. It may not be caused by `dp` has a smaller lengt. In some special cases, users may implement a`MapDataPipe` with fancy indexing mechanism like using string or starting from 1000. So, I think the error should simply be like `index is out of range`.

NivekT replied:
Good catch - fixed.

ejguan reviewed (approved):
LGTM with one comment

NivekT committed:
Update on "[DataPipe] adding ZipperMapDataPipe"


Part of #57031

cc @VitalyFedyunin ejguan @NivekT

Differential Revision: [D32263058](https://our.internmc.facebook.com/intern/diff/D32263058)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32263058).
